File: 백준/Silver/1012. 유기농 배추/유기농 배추.py
import sys
input = sys.stdin.readline
sys.setrecursionlimit(10**6)

dx = [0, 0, -1, 1]
dy = [-1, 1, 0, 0]

def dfs(x, y):
    if x < 0 or x >= width or y < 0 or y >= length:
        return
    if field[y][x] == 1:
        field[y][x] = 0
        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]
            dfs(nx, ny)

num = int(input())
while num:
    width, length, cab_num = map(int, input().split())
    field = [[0 for _ in range(width)] for _ in range(length)]
    for _ in range(cab_num):
        index_x, index_y = map(int, input().split())
        field[index_y][index_x] = 1
    count = 0
    for i in range(length):
        for j in range(width):
            if field[i][j] == 1:
                dfs(j, i)
                count += 1
    print(count)
    num -= 1

File: 백준/Silver/1012. 유기농 배추/유기농 배추.c
#include <stdio.h>
#include <string.h>

#define MAX_SIZE 50

int field[MAX_SIZE][MAX_SIZE];
int width, length;

int dx[] = {0, 0, -1, 1};
int dy[] = {-1, 1, 0, 0};

void dfs(int x, int y) {
    if (x < 0 || x >= width || y < 0 || y >= length) {
        return;
    }

    if (field[y][x] == 1) {
        field[y][x] = 0;
        for (int i = 0; i < 4; i++) {
            int nx = x + dx[i];
            int ny = y + dy[i];
            dfs(nx, ny);
        }
    }
}

int main() {
    int num;
    scanf("%d", &num);

    while (num--) {
        int cab_num;
        scanf("%d %d %d", &width, &length, &cab_num);

        memset(field, 0, sizeof(field)); // 0으로 초기화

        for (int i = 0; i < cab_num; i++) {
            int index_x, index_y;
            scanf("%d %d", &index_x, &index_y);
            field[index_y][index_x] = 1;
        }

        int count = 0;
        for (int i = 0; i < length; i++) {
            for (int j = 0; j < width; j++) {
                if (field[i][j] == 1) {
                    dfs(j, i);
                    count++;
                }
            }
        }
        printf("%d\n", count);
    }

    return 0;
}